### PR TITLE
Ensure instances required for menus are present

### PR DESCRIPTION
**Pull request by esoergel**
## Product Description
Addresses this error: https://dimagi.sentry.io/discover/commcarehq:740939177fc048389e87590d059eee61/

> [WebApps] The instance "casedb" in expression "instance(casedb)/casedb/case..." used by "/" does not exist in the form. Please correct your form or...

## Technical Summary

https://dimagi-dev.atlassian.net/browse/SUPPORT-17147
https://dimagi-dev.atlassian.net/browse/SUPPORT-17186

Those issues were caused by https://github.com/dimagi/commcare-hq/pull/33037

This PR reverts part of the change from this commit: https://github.com/dimagi/commcare-hq/commit/ac1a51512b372bfdcb4a9ff52dbba43eeec53189
That commit changed two things:

* entry-associated instances were no longer propagated to all entries in the menu
* menu-associated instances were no longer included in entries from other menus with the same menu ID

That latter change was inadvertent - I didn't consider that menus can have multiple IDs, and that from the commcare-core perspective, this makes them essentially the same menu.

Anyways, this commit groups instances required by menus by menu ID, then distributes to all forms in menus with that ID.  The test fails without the associated suite generation change.

### Safety story
Automated tests, primarily.  We didn't have any tests exercising the functionality that broke, and it wasn't caught by QA.  I did add that test in this PR.  Since this is a revert to previous behavior (if not previous code), I think this is at least less likely to break things than the current version of the code.

### Rollback instructions
- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Review by shubham1g5 — APPROVED** [review]

File: corehq/apps/app_manager/tests/test_suite_instances.py [hunk 1]
```diff
@@ -260,17 +260,27 @@ def test_module_filter_instances_on_all_forms(self):
...
+        self.m0, self.m0f0 = factory.new_basic_module('m0', 'case1')
+        self.m0.put_in_root = True
+
+        self.m1, self.m1f0 = factory.new_basic_module('m1', 'case1')
+        self.m1.module_filter = "instance('locations')/locations/"
+        self.m1.put_in_root = True
+
+        suite = factory.app.create_suite()
+        instance_xml = "<partial><instance id='locations' src='jr://fixture/locations' /></partial>"
+        self.assertXmlPartialEqual(instance_xml, suite, "entry/command[@id='m0-f0']/..instance")
```
Comment (orangejenny) [inline corehq/apps/app_manager/tests/test_suite_instances.py:282]: I don't know the `..` syntax, what's that? I'm also uncertain what the `..` in `entry/command[@id='m0-f1']/../instance` is for - instances are direct children of commands, so is it needed?

**Review by orangejenny — APPROVED** [review]
Question about test syntax, but code looks good.

File: corehq/apps/app_manager/tests/test_suite_instances.py (hunk 1 above)
Comment (esoergel) [reply corehq/apps/app_manager/tests/test_suite_instances.py:282]: The `..` refers to the parent level in xpath, similar to `..` in unix filesystems.  I pulled the syntax from some other test files (I see it in `test_suite_case_tiles` and `test_shadow_forms`).  `instance` is actually a sibling of `command`, here's what this `entry` block looks like:

```xml
<entry>
  <command id="m1-f0">
    <text>
      <locale id="forms.m1f0"/>
    </text>
  </command>
  <instance id="locations" src="jr://fixture/locations"/>
</entry>
```
It's unfortunate that the `m1-f0` label is on the `command`, not the `entry`, as I don't think there's another good way to identify the entry you want (index, I guess, but that's not great).

I do notice I did `..instance` here and `../instance` above - both seem to work, but the latter seems more correct, so I'll update that.

Comment (orangejenny) [reply corehq/apps/app_manager/tests/test_suite_instances.py]: oh, I see, I must have misread the suite. Thanks for laying out an example.

**Pull request merged**